<a href="https://colab.research.google.com/github/ReneSRuiz/Mineria_de_datos/blob/main/AvancePIA_II_001_11.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Equipo:** 11  
**Grupo:** 001  
**Frecuencia:** lunes  
**Integrantes:**

 *   García Alanís Alejandro Uriel 1886968
 *   Estrada Dávila Jorge Eduardo 1741943
 *   Paz Reyes Néstor Misael 1559508
 *   González García Elisa 1858207
 *   Sobrevilla Ruiz Rene 1941452  

**BASE DE DATOS PRINCIPAL**  
**1) Título de la base de datos**  
En nombre oficial de la base de datos es "Lichess September 2020 data", la cual se puede encontrar por el siguiente enlace: https://www.kaggle.com/noobiedatascientist/lichess-september-2020-data?select=Sept_20_analysis.csv  
**2) Introducción general**  
El ajedrez es un juego con muchos años de antigüedad, sencillo de entender pero con una dificultad estratégica muy interesante. Se cree que los grandes reyes y estrategas militares veían al ajedrez como una forma de estudiar a sus enemigos y mejorar como líderes. El ajedrez a tenido momentos importantes en la historia humana, desde juegos amistosos para generar alianzas entre reyes como el fin de épocas importantes como lo fué la primera victoria de un americano contra la URSS, país que dominaba este deporte en tiempos de guerra fría, significando una victoria moral para EUA.  
El ajedrez era un deporte de nicho, popular entre gente que buscaba un juego estratégico y gente que ama el análisis tan complejo de este. Pero en tiempos modernos, con estrenos de series muy famosas, ejemplo Gambito de Dama, este juego ha alcanzado una popularuidad muy alta. Ahora, no solamente las personas "inteligentes" lo juegan, sino que la persona común se ha animado a aprender de este juego tan impresionante y educativo.  
**3) Descripción de los datos**  
Los datos de la base seleccionada estan en una tabla. En total, son más de 3 millones de filas, y 40 columnas. Cada fila representa un enfrentamiento entre 2 personas y las columnas nos revelan varios datos, como el tipo de juego, el "elo" de los jugadores, la fecha del juego, resultados, momentos importantes, nombre de aperturas, movimientos y errores de ambos, entre otros. Hay datos tanto enteros, flotantes y objetos (cadenas de caracteres y fechas).  
**4) Justificación del uso de datos.**   
Creemos que la base de datos es muy completa, ya que tiene detallado cada movimiento de los jugadores, como el estado de la partida que estos jugaron. Además, la información está muy completa, sin muchos datos nulos que puedan afectar nuestro trabajo.
Hablando de los beneficios, al estar en su mayoria completos, creemos que podemos obtener resultados significativos y que nos ayuden a solucionar el problema planteado y analizar este juego de una manera más significativa. Es verdad que al ser muchos datos, se tardará en realizar los procesos programados, pero con paciencia, estos pueden ser muy reveladores.  
**5) Objetivo final.**  
Nuestro objetivo principal es poder clasificar las mejores aperturas del ajedrez, dependiento del "elo" que tengan los jugadores y la eficacia de estas aperturas para la creación de entrategias ganadoras.  
Creemos que, con estos datos, podemos crear la clasificación antes mencionada y ayudar a las personas nuevas, o un poco mas experimentadas, a saber cual apertura funciona mejor dependiendo de su desempeño como de gustos, y asi, poder estudiar la apertura perfecta o cambiar su estrategia para que esta se ajuste a su estilo.  
**6) Mencionar preguntas de interés.**  
Hay varias cosas interesantes que podemos indagar sobre esta base, como lo pueden ser:  
* ¿Se juegan más aperturas clásicas o modernas?  
* ¿El tiempo promedio de juego aumenta conforme el rival?, es decir, ¿Mientras mas "elo" tienen los rivales, mas largas son estas partidas?
* ¿Son igual de efectivas las aperturas populares entre la gente de "elo bajo" en los "elos altos"?  
* ¿Las aperturas favoritas de los "elos altos" son efectivas en los "elos bajos"?
* ¿Que tantas jugadas inadecuadas, meteduiras de patas o volteos de juego tiene un jugador de "low elo" comparado con los de mayor "elo"?  
**LIMPIEZA DE DATOS**  
Iniciaremos cargando los datos y mostrándolos sin manipular, además de que cargaremos las librerias necesarias para este apartado y los futuros.

In [8]:
#importando las librerias
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [ ]:
#Importar la base de datos al google colab
from google.colab import files
files.upload()

Saving Sept_20_analysis.csv to Sept_20_analysis.csv


In [9]:
#Nuestros datos a trabajar son
df = pd.read_csv("Sept_20_analysis.csv")
#Visualizando tabla
df.head(5)

GAME  BlackElo  ...  Game_flips Game_flips_ts
0    11      1143  ...           8             0
1    14      1504  ...           6             0
2    29      1933  ...           5             0
3    40      1710  ...           8             1
4    55      1598  ...           2             0

[5 rows x 40 columns]

In [10]:
#Vemos si hay datos nulos
df.isnull().sum()

GAME                        0
BlackElo                    0
BlackRatingDiff         13036
Date                        0
ECO                         0
Event                       0
Opening                     0
Result                      0
Site                        0
Termination                 0
TimeControl                 0
UTCTime                     0
WhiteElo                    0
WhiteRatingDiff         13057
Black_elo_category          0
White_elo_category          0
starting_time           10344
increment               10344
Game_type                   0
Total_moves                 0
Black_blunders              0
White_blunders              0
Black_mistakes              0
White_mistakes              0
Black_inaccuracies          0
White_inaccuracies          0
Black_inferior_moves        0
White_inferior_moves        0
Black_ts_moves              0
White_ts_moves              0
Black_ts_blunders           0
White_ts_blunders           0
Black_ts_mistakes           0
White_ts_m

Vemos que esta base de datos cuenta con bastante información interesante, pero en nuestro caso, nos enfocamos en movimientos, aperturas y los mismos errores de los jugadores, por lo que columnas como la diferencia de rating y la de los tiempos de juego no son de interés, además de que cuentan con datos nulos y nos quitaríamos un problema. Además de estas columnas, hay otras mas que no tienen mucha información, como el ID del juego o la fecha del mismo, por lo que igual se considerarian a eliminar. Entonces, con el criterio de eliminar columnas que no contengan información relevante, que no aportan a nuestros objetivos y que tengan datos nulos, procedemos a eliminar estas columnas, además de renombrar las que se conservan por temas de practicidad en el futuro. Además, realizaremos una copia de los datos en un CSV para alivianar el peso de la base de datos, ya que la original cuenta con bastantes datos, y aminorar el peso del mismo nos ayudará en un futuro.  
Para mayor profundidad, estas son las razones de las eliminaciones de las columnas:  
* **GAME, Date, Site:** Tanto fechas como ID del juego y el donde encontrarlo no son importantes.  
* **BlackRatingDiff, WhiteRatingDiff:** Pueden sacarse a mano, además de que no aporta a nuestros objetivos.  
* **starting_time, increment, TimeControl:** Nos enfocamos a los movimientos, no al tiempo de juego de los jugadores.  
* **UTCTime, ECO:** No revela información importante para nuestros objetivos.  
* **Black_inferior_moves, White_inferior_moves, Black_ts_moves, White_ts_moves, Black_ts_blunders, White_ts_blunders, Black_ts_mistakes, White_ts_mistake, Black_long_moves, White_long_moves, Black_bad_long_moves, White_bad_long_moves, Game_flips_ts:** Nos enfocamos en los errores estándares.
* **Event:** Para eso está la columna de Game_Type.

In [11]:
#Eliminación de columnas
df2=df.drop({'GAME','BlackRatingDiff','Date','ECO','Site','UTCTime','WhiteRatingDiff','starting_time','increment',
                 'Black_inferior_moves', 'White_inferior_moves', 'Black_ts_moves','White_ts_moves','Black_ts_blunders', 
                 'White_ts_blunders','TimeControl','Black_ts_mistakes', 'White_ts_mistake', 'Black_long_moves','White_long_moves',
                 'Black_bad_long_moves','Event', 'White_bad_long_moves','Game_flips_ts'},axis = 1 )
#Cambiaremos los nombres de inglés a español
df2=df2.rename(columns = {'Opening':'Apertura','Result':'Resultado','Termination':'Terminacion',
                                    'Black_elo_category':'Categoria_Negras',
                                    'White_elo_category':'Categoria_Blancas','Game_type':'Tipo_juego',
                                    'Total_moves':'Movimientos','Black_blunders':'Negras_Metedura_de_patas',
                                    'White_blunders':'Blancas_Metedura_de_patas','Black_mistakes':'Negras_Errores',
                                    'White_mistakes':'Blancas_Errores','Black_inaccuracies':'Negras_Imprecision',
                                    'White_inaccuracies':'Blancas_Imprecision','Game_flips':'Juego_Vueltas',})
#Guardar el archivo en un csv nuevo para futuros trabajos
df2.to_csv("Base_Principal_Modificada.csv", index=False)
df2.head(5)

BlackElo                  Apertura  ... Blancas_Imprecision Juego_Vueltas
0      1143              Bird Opening  ...                   1             8
1      1504              Réti Opening  ...                   5             6
2      1933          Philidor Defense  ...                   8             5
3      1710  Sicilian Defense: Closed  ...                   4             8
4      1598          Alekhine Defense  ...                   2             2

[5 rows x 16 columns]

El nombre o datos en cadenas de caracteres se quedarán en su idioma original (Inglés), debido a ser un lenguaje universal estandar dentro del ajedrez. A pesar de estar en su idioma, dentro de los análisis de traducirán según su conveniencia.  
**ESTADISTICA BÁSICA**  
Ahora realizaremos, de forma descriptiva, nuestra base de datos. Primero hablaremos de las columnas que tengan datos en forma de cadenas de caracteres, para después seguir con los numéricos, los cuales tendrán un análisis más profundo.

In [12]:
df2['Apertura'].describe()

count                             3739909
unique                               2790
top       Queen's Pawn Game: Mason Attack
freq                                76214
Name: Apertura, dtype: object

Vemos que en cuestión de aperturas, se registraron al menos 2,790 aperturas diferentes, de la cual, la mas jugada es "Queen's Pawn Game: Mason Attack".

In [24]:
df2['Resultado'].describe()

count     3739909
unique          4
top           1-0
freq      1863894
Name: Resultado, dtype: object


In [26]:
df2['Resultado'].unique()

array(['0-1', '1-0', '1/2-1/2', '*'], dtype=object)

En cuestión de los resultados, sabemos que estos son declarar un ganador, ya sea 0-1 (Gana negras) o 1-0 (Gana blancas), pero igualmente hay empates (1/2-1/2) y abandonos o juegos con fallos ('*'), y por lo que vemos, los resultados mencionan que las blancas ganan con mayor frecuencia.

In [22]:
df2['Terminacion'].describe()

count     3739909
unique          4
top        Normal
freq      2869458
Name: Terminacion, dtype: object

In [23]:
df2['Terminacion'].unique()

array(['Time forfeit', 'Normal', 'Rules infraction', 'Abandoned'],
      dtype=object)

De la mano con los resultados, la forma en como estos son determinados son importantes. Vemos que pueden ser por tiempo, de forma normal, por infracciones o abandonos, y que lo más frecuente es que se determine un ganador o empate de forma normal, es decir, con algún "JAQUE MATE", que algún jugador se rinda o que ambos pacten el empate.

In [28]:
print(df2['Categoria_Blancas'].describe(),'\n')
print(df2['Categoria_Negras'].describe())

count        3739909
unique             3
top       Low rating
freq         2993539
Name: Categoria_Blancas, dtype: object 

count        3739909
unique             3
top       Low rating
freq         2991089
Name: Categoria_Negras, dtype: object


In [32]:
df2['Categoria_Blancas'].unique()

array(['Low rating', 'High rating', 'GM rating'], dtype=object)

Recordando que los jugadores tienen categorías, estas van desde el "LOW RATING", "HIGH RATING" y "GM RATING", y vemos que, tanto jugadores de blancas como negras, se situan en el "LOW RATING", lo cual es de esperarse, ya que subir a los demas ratings requieren mayor expertis en el juego, y seguramente muchos de los jugadores lo hacen por hobbie.

In [30]:
df2['Tipo_juego'].describe()

count     3739909
unique          5
top         Blitz
freq      1812120
Name: Tipo_juego, dtype: object

In [31]:
df2['Tipo_juego'].unique()

array(['Blitz', 'Rapid', 'Bullet', 'Classical', 'Correspondence'],
      dtype=object)

Igualmente vemos que el modo/tipo de juego favorito es erl "BLITZ", esto puede deberse a que este modo es rápido, y más amigable o interesante para los jugadores.  
Ahora pasaremos a los datos numéricos:

In [36]:
print(df2['BlackElo'].describe(), '\n')
print('Varianza: ', df2['BlackElo'].var(), '\n')
print('Desviación estandar: ', df2['BlackElo'].std(), '\n')
print("Simetría de los datos:", df2['BlackElo'].skew(), "\n")
print("Curtosis", df2['BlackElo'].kurt())

count    3.739909e+06
mean     1.602300e+03
std      3.435793e+02
min      6.000000e+02
25%      1.354000e+03
50%      1.593000e+03
75%      1.841000e+03
max      3.957000e+03
Name: BlackElo, dtype: float64 

Varianza:  118046.76030488501 

Desviación estandar:  343.57933626003324 

Simetría de los datos: 0.1607547371835947 

Curtosis -0.24272112523246037


Vemos que el jugador de negras con menor elo es de 600 y el de mayor elo es de 3957, además la desviación estandar indica que hay mucha dispersión, por lo que los jugadores son muy variados. Además el coeficiente de simetría podría indicar que estos datos tienen algún tipo de simetría, por lo que podrían estar bien repartidos los jugadores, esto reforzado por la curtosis, que nos dice que muchos jugadores no se alejan de la media.

In [37]:
print(df2['WhiteElo'].describe(), '\n')
print('Varianza: ', df2['WhiteElo'].var(), '\n')
print('Desviación estandar: ', df2['WhiteElo'].std(), '\n')
print("Simetría de los datos:", df2['WhiteElo'].skew(), "\n")
print("Curtosis", df2['WhiteElo'].kurt())

count    3.739909e+06
mean     1.601512e+03
std      3.435898e+02
min      6.000000e+02
25%      1.353000e+03
50%      1.592000e+03
75%      1.840000e+03
max      3.958000e+03
Name: WhiteElo, dtype: float64 

Varianza:  118053.94356861341 

Desviación estandar:  343.5897896745673 

Simetría de los datos: 0.15853401134895811 

Curtosis -0.24096420125435936


Vemos que el jugador de blancas con menor elo es de 600 y el de mayor elo es de 3958, además la desviación estandar indica que hay mucha dispersión, por lo que los jugadores son muy variados. Además el coeficiente de simetría podría indicar que estos datos tienen algún tipo de simetría, al igual que con los jugadores de negras, por lo que podrían estar bien repartidos los jugadores, esto reforzado por la curtosis, que nos dice que muchos jugadores no se alejan de la media.

In [40]:
print(df2['Movimientos'].describe(), '\n')
print('Varianza: ', df2['Movimientos'].var(), '\n')
print('Desviación estandar: ', df2['Movimientos'].std(), '\n')
print("Simetría de los datos:", df2['Movimientos'].skew(), "\n")
print("Curtosis", df2['Movimientos'].kurt())

count    3.739909e+06
mean     6.402238e+01
std      2.672077e+01
min      1.000000e+00
25%      4.600000e+01
50%      5.900000e+01
75%      7.700000e+01
max      3.530000e+02
Name: Movimientos, dtype: float64 

Varianza:  713.9995436050986 

Desviación estandar:  26.720769891698453 

Simetría de los datos: 1.2389245621873834 

Curtosis 2.53195619347439


Vemos que hay juegos que solamente hubo 1 movimiento, popsiblemente ganando blancas por abandono, y juegos más largos de hasta 353 movimientos. Además, el coeficiente de simetría indica que hay asimetría positiva, por lo que los juegos con "muchos" movimientos son los menos frecuentes, además, la curtosis indica que muchos de los juegos tienen muvimientos alejados de la media, el cual es 64 movimientos por fuejo, por lo que podríamos esperar que hay tanto juegos que terminan muy rápido como juegos que se alargan, habiendo variedad en los mismos.

In [41]:
print(df2['Negras_Metedura_de_patas'].describe(), '\n')
print('Varianza: ', df2['Negras_Metedura_de_patas'].var(), '\n')
print('Desviación estandar: ', df2['Negras_Metedura_de_patas'].std(), '\n')
print("Simetría de los datos:", df2['Negras_Metedura_de_patas'].skew(), "\n")
print("Curtosis", df2['Negras_Metedura_de_patas'].kurt(),'\n\n\n')
print(df2['Blancas_Metedura_de_patas'].describe(), '\n')
print('Varianza: ', df2['Blancas_Metedura_de_patas'].var(), '\n')
print('Desviación estandar: ', df2['Blancas_Metedura_de_patas'].std(), '\n')
print("Simetría de los datos:", df2['Blancas_Metedura_de_patas'].skew(), "\n")
print("Curtosis", df2['Blancas_Metedura_de_patas'].kurt(),'\n\n\n')

count    3.739909e+06
mean     2.042054e+00
std      1.961390e+00
min      0.000000e+00
25%      1.000000e+00
50%      2.000000e+00
75%      3.000000e+00
max      3.200000e+01
Name: Negras_Metedura_de_patas, dtype: float64 

Varianza:  3.847050449537979 

Desviación estandar:  1.9613899279689337 

Simetría de los datos: 1.674330907406602 

Curtosis 4.876140396669593 



count    3.739909e+06
mean     2.021308e+00
std      1.974633e+00
min      0.000000e+00
25%      1.000000e+00
50%      2.000000e+00
75%      3.000000e+00
max      3.600000e+01
Name: Blancas_Metedura_de_patas, dtype: float64 

Varianza:  3.899174237081576 

Desviación estandar:  1.974632684091291 

Simetría de los datos: 1.675853812170047 

Curtosis 4.8391136230259 





En cuestión de las "Meteduras de patas", estas parecen comportarse de manera similar tanto para blancas como negras. Ambos tienen disviaciones muy pequeñas y mínimos de 0 "Meteduras de patas", pero el máximo de "meteduras de patas" de negras y blancas son de 39 y 36 respectivamente. Además, en ambos casos cuentan con asimetría positiva y curtosis demasiado altas, por lo que los juegos, por lo general, deben tener pocas "meteduras de pata", miestras que los juegos que tienen grandes cantidades, como 36 o 39 "meteduras de patas", son juegos muy atípicos.

In [42]:
print(df2['Negras_Errores'].describe(), '\n')
print('Varianza: ', df2['Negras_Errores'].var(), '\n')
print('Desviación estandar: ', df2['Negras_Errores'].std(), '\n')
print("Simetría de los datos:", df2['Negras_Errores'].skew(), "\n")
print("Curtosis", df2['Negras_Errores'].kurt(),'\n\n\n')
print(df2['Blancas_Errores'].describe(), '\n')
print('Varianza: ', df2['Blancas_Errores'].var(), '\n')
print('Desviación estandar: ', df2['Blancas_Errores'].std(), '\n')
print("Simetría de los datos:", df2['Blancas_Errores'].skew(), "\n")
print("Curtosis", df2['Blancas_Errores'].kurt(),'\n\n\n')

count    3.739909e+06
mean     3.081397e+00
std      2.334158e+00
min      0.000000e+00
25%      1.000000e+00
50%      3.000000e+00
75%      4.000000e+00
max      2.800000e+01
Name: Negras_Errores, dtype: float64 

Varianza:  5.448291348161977 

Desviación estandar:  2.334157524281936 

Simetría de los datos: 0.9773190867942845 

Curtosis 1.2462225923675274 



count    3.739909e+06
mean     3.070554e+00
std      2.347220e+00
min      0.000000e+00
25%      1.000000e+00
50%      3.000000e+00
75%      4.000000e+00
max      3.900000e+01
Name: Blancas_Errores, dtype: float64 

Varianza:  5.509439656787787 

Desviación estandar:  2.3472195587093654 

Simetría de los datos: 0.9782795202378635 

Curtosis 1.231731956408097 





De igual manera que con las "Meteduras de patas", los "errores" estas parecen comportarse de manera similar tanto para blancas como negras. Ambos tienen disviaciones muy pequeñas y mínimos de 0 "errores" , pero el máximo de "errores"  de negras y blancas son de 39 y 40 respectivamente. Además, en ambos casos cuentan con asimetría positiva y curtosis ligeramente elevadas, por lo que los juegos, por lo general, deben tener pocos "errores" , miestras que los juegos que tienen grandes cantidades, como 39 o 40 "errores" , son juegos muy atípicos.

In [43]:
print(df2['Negras_Imprecision'].describe(), '\n')
print('Varianza: ', df2['Negras_Imprecision'].var(), '\n')
print('Desviación estandar: ', df2['Negras_Imprecision'].std(), '\n')
print("Simetría de los datos:", df2['Negras_Imprecision'].skew(), "\n")
print("Curtosis", df2['Negras_Imprecision'].kurt(),'\n\n\n')
print(df2['Blancas_Imprecision'].describe(), '\n')
print('Varianza: ', df2['Blancas_Imprecision'].var(), '\n')
print('Desviación estandar: ', df2['Blancas_Imprecision'].std(), '\n')
print("Simetría de los datos:", df2['Blancas_Imprecision'].skew(), "\n")
print("Curtosis", df2['Blancas_Imprecision'].kurt(),'\n\n\n')

count    3.739909e+06
mean     2.893094e+00
std      2.035020e+00
min      0.000000e+00
25%      1.000000e+00
50%      3.000000e+00
75%      4.000000e+00
max      2.300000e+01
Name: Negras_Imprecision, dtype: float64 

Varianza:  4.141308381840579 

Desviación estandar:  2.0350204868355943 

Simetría de los datos: 0.9051260324627162 

Curtosis 1.1415605305994512 



count    3.739909e+06
mean     2.794539e+00
std      2.039361e+00
min      0.000000e+00
25%      1.000000e+00
50%      2.000000e+00
75%      4.000000e+00
max      2.700000e+01
Name: Blancas_Imprecision, dtype: float64 

Varianza:  4.158993705566677 

Desviación estandar:  2.039361102298138 

Simetría de los datos: 0.9256483983264125 

Curtosis 1.1568794000626546 





Al igual que con los anteriores movimientos, las "impresiciones" parecen comportarse de manera similar tanto para blancas como negras. Ambos tienen disviaciones muy pequeñas y mínimos de 0 "impresiciones", pero el máximo de "impresiciones"  de negras y blancas son de 23 y 27 respectivamente. Además, en ambos casos cuentan con asimetría positiva y curtosis ligeramente elevadas, por lo que los juegos, por lo general, deben tener pocos "impresiciones", miestras que los juegos que tienen grandes cantidades, como 39 o 40 "impresiciones", son juegos muy atípicos.

In [45]:
print(df2['Juego_Vueltas'].describe(), '\n')
print('Varianza: ', df2['Juego_Vueltas'].var(), '\n')
print('Desviación estandar: ', df2['Juego_Vueltas'].std(), '\n')
print("Simetría de los datos:", df2['Juego_Vueltas'].skew(), "\n")
print("Curtosis", df2['Juego_Vueltas'].kurt())

count    3.739909e+06
mean     6.251911e+00
std      4.739733e+00
min      0.000000e+00
25%      3.000000e+00
50%      5.000000e+00
75%      9.000000e+00
max      8.300000e+01
Name: Juego_Vueltas, dtype: float64 

Varianza:  22.465073515142542 

Desviación estandar:  4.739733485665892 

Simetría de los datos: 1.2683948770567326 

Curtosis 2.270590063811904


Por último, analizaremos las veces que el juego cambió de lado, es decir, cuando un jugador que tenía las de perder, logra tomar ventaja. En este caso, igualmente hay juegos dominados por 1 bando, es decir, 0 cambios de ventaja, y hay juegos donde cambiaron papeles hasta 83 veces. Vemos que la desviación estandar es poca, que hay asimetría positiva y la curtosis es muy alta, por lo que da indicios de que los juegos con varios cambios de ventajas, ya sea mas de 60 o 70 veces, son realmente atípicos, y que la mayoría puede que ronde entre unas 6 veces el cambio de ventajas o de posiciones ganadoras.  
Con esto, damos por terminado la parte descriptiva, ahora veremos gráficas diversas que nos pueden aportar una mejor visualización de la información.  
**GRÁFICAS**